### PR TITLE
Setup private extensions as vcs

### PR DESCRIPTION
**Pull request by msz-rai**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[submodule "udp"]
+	path = extensions/udp
+	url = git@github.com:RobotecAI/RGL-extension-udp.git
+	branch = feature/init-extension
```
Comment (msz-rai) [inline .gitmodules:4]: The branch is to be changed after merging PR in the extension repo.

**Review by PawelLiberadzki — APPROVED** [review]

File: src/TapeCore.hpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	static void tape_node_points_compact(const YAML::Node& yamlNode, TapeState& tapeState);
+	static void tape_node_points_spatial_merge(const YAML::Node& yamlNode, TapeState& tapeState);
+	static void tape_node_points_temporal_merge(const YAML::Node& yamlNode, TapeState& tapeState);
+	static void tape_node_points_from_array(const YAML::Node& yamlNode, TapeState& tapeState);
+	static void tape_node_gaussian_noise_angular_ray(const YAML::Node& yamlNode, TapeState& tapeState);
+	static void tape_node_gaussian_noise_angular_hitpoint(const YAML::Node& yamlNode, TapeState& tapeState);
+	static void tape_node_gaussian_noise_distance(const YAML::Node& yamlNode, TapeState& tapeState);
+
+	// Called once in the translation unit
+	static inline bool autoExtendTapeFunctions = std::invoke([]() {
```
Comment (prybicki) [inline src/TapeCore.hpp:64]: :1st_place_medal:

File: src/api/apiCore.cpp [hunk 3]
```diff
@@ -62,7 +64,7 @@ RGL_API rgl_status_t rgl_get_version_info(int32_t* out_major, int32_t* out_minor
 	return status;
 }
 
-void TapePlayer::tape_get_version_info(const YAML::Node& yamlNode)
+void TapeCore::tape_get_version_info(const YAML::Node& yamlNode, TapeState& tapeState)
```
Comment (prybicki) [inline src/api/apiCore.cpp:67]: Minor: consider `tapeState` -> `state`. It's rather obvious from the context, and `playbackState` might be lengthy.

File: src/api/apiCore.cpp [hunk 4]
```diff
@@ -153,12 +155,13 @@ RGL_API rgl_status_t rgl_cleanup(void)
...
+void TapeCore::tape_cleanup(const YAML::Node& yamlNode, TapeState& tapeState)
 {
 	rgl_cleanup();
-	tapeMeshes.clear();
-	tapeEntities.clear();
-	tapeNodes.clear();
+	tapeState.meshes.clear();
+	tapeState.entities.clear();
+	tapeState.nodes.clear();
+	tapeState.textures.clear();
```
Comment (prybicki) [inline src/api/apiCore.cpp:164]: :1st_place_medal:

**Review by prybicki — CHANGES_REQUESTED** [review]
Good changes. Left a few suggestions to consider.

**Review by prybicki — APPROVED** [review]
Re-review to fix accidental `request changes` in the previous review.

File: .gitignore [hunk 5]
```diff
@@ -40,3 +40,4 @@ cmake-build-debug
 cmake-build-release
 build
 external/vcpkg
+extensions/udp
```
Comment (prybicki) [inline .gitignore:43]: :1st_place_medal:

File: src/api/apiCore.cpp (hunk 4 above)
Comment (msz-rai) [reply src/api/apiCore.cpp:164]: I have created a method `clear` in `PlaybackState` to remember about clearing when a new TapeAPIObject will be added. When `textures` were added, we forgot to clear them in `tape_cleanup` (`rgl_cleanup` playback function).

File: src/Tape.hpp [hunk 6]
```diff
@@ -180,11 +185,26 @@ class TapeRecorder
...
+	std::unordered_map<TapeAPIObjectID, rgl_texture_t> textures;
+	std::unordered_map<TapeAPIObjectID, rgl_node_t> nodes;
+	uint8_t* fileMmap{};
+	size_t mmapSize{};
+};
+
+// Signature of tape function corresponding to the API function
+using TapeFunction = std::function<void(const YAML::Node&, TapeState&)>;
+
 struct TapePlayer
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply src/Tape.hpp]: Done

File: src/Tape.hpp [hunk 7]
```diff
@@ -195,12 +195,27 @@ struct PlaybackState
...
 	}
 
+	template<typename T>
+	T* getPtr(const YAML::Node& offsetYamlNode)
+	{
+		assert(fileMmap != nullptr);
+		auto offset = offsetYamlNode.as<size_t>();
+		if (offset > mmapSize) {
+			throw std::runtime_error(fmt::format("Tape binary offset ({}) out of range ({})", offset, mmapSize));
+		}
```
Comment (prybicki) [inline src/Tape.hpp]: Isn't it supposed to be `>=`? I'd expect `offset` can be [0, mmapSize)

**Review by prybicki — APPROVED** [review]
LGTM, but see the comment - potential bug.

File: src/Tape.hpp (hunk 7 above)
Comment (msz-rai) [reply src/Tape.hpp]: You are right.
More precisely, shouldn't it be `[0, mmapSize - sizeof(T)]`?

**Pull request merged**